### ISSUE TITLE
Training yolact is too slow

### ISSUE THREAD

**Issue by zhiweihu1103** [open-mmlab/mmdetection#4628]
I use yolact to train on my own training set, but the training speed and FPS are slower than Mask RCNN. Can anyone help? Below is my configuration file yolact_r50_1x8.py. I only change the **num_classes**, and datasets path. On Tesla P100, the FPS of Mask RCNN is 22.7, but YOLACT is 14. I use the mmdetection/tree/master/tools/analysis_tools/benchmark.py to test the FPS. Besides, our model size is 287M, the [official size](https://openmmlab.oss-cn-hangzhou.aliyuncs.com/mmdetection/v2.0/yolact/yolact_r50_1x8_coco_20200908-f38d58df.pth) is 135M. I'm not sure what the reason is？
When I execute **nvidia-smi**, I find the **volatile gpu-util** also too low but **Memory-Usage** is high.
```
_base_ = '../_base_/default_runtime.py'
img_size = 550
model = dict(
    type='YOLACT',
    pretrained='torchvision://resnet50',
    backbone=dict(
        type='ResNet',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=-1,  # do not freeze stem
        norm_cfg=dict(type='BN', requires_grad=True),
        norm_eval=False,  # update the statistics of bn
        zero_init_residual=False,
        style='pytorch'),
    neck=dict(
        type='FPN',
        in_channels=[256, 512, 1024, 2048],
        out_channels=256,
        start_level=1,
        add_extra_convs='on_input',
        num_outs=5,
        upsample_cfg=dict(mode='bilinear')),
    bbox_head=dict(
        type='YOLACTHead',
        num_classes=1,
        in_channels=256,
        feat_channels=256,
        anchor_generator=dict(
            type='AnchorGenerator',
            octave_base_scale=3,
            scales_per_octave=1,
            base_sizes=[8, 16, 32, 64, 128],
            ratios=[0.5, 1.0, 2.0],
            strides=[550.0 / x for x in [69, 35, 18, 9, 5]],
            centers=[(550 * 0.5 / x, 550 * 0.5 / x)
                     for x in [69, 35, 18, 9, 5]]),
        bbox_coder=dict(
            type='DeltaXYWHBBoxCoder',
            target_means=[.0, .0, .0, .0],
            target_stds=[0.1, 0.1, 0.2, 0.2]),
        loss_cls=dict(
            type='CrossEntropyLoss',
            use_sigmoid=False,
            reduction='none',
            loss_weight=1.0),
        loss_bbox=dict(type='SmoothL1Loss', beta=1.0, loss_weight=1.5),
        num_head_convs=1,
        num_protos=32,
        use_ohem=True),
    mask_head=dict(
        type='YOLACTProtonet',
        in_channels=256,
        num_protos=32,
        num_classes=1,
        max_masks_to_train=100,
        loss_mask_weight=6.125),
    segm_head=dict(
        type='YOLACTSegmHead',
        num_classes=1,
        in_channels=256,
        loss_segm=dict(
            type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0)))
train_cfg = dict(
    assigner=dict(
        type='MaxIoUAssigner',
        pos_iou_thr=0.5,
        neg_iou_thr=0.4,
        min_pos_iou=0.,
        ignore_iof_thr=-1,
        gt_max_assign_all=False),
    # smoothl1_beta=1.,
    allowed_border=-1,
    pos_weight=-1,
    neg_pos_ratio=3,
    debug=False)
test_cfg = dict(
    nms_pre=1000,
    min_bbox_size=0,
    score_thr=0.05,
    iou_thr=0.5,
    top_k=200,
    max_per_img=100)
dataset_type = 'CocoDataset'
data_root = 'data/pig/'
img_norm_cfg = dict(
    mean=[123.68, 116.78, 103.94], std=[58.40, 57.12, 57.38], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile', to_float32=True),
    dict(type='LoadAnnotations', with_bbox=True, with_mask=True),
    dict(type='FilterAnnotations', min_gt_bbox_wh=(4.0, 4.0)),
    dict(
        type='PhotoMetricDistortion',
        brightness_delta=32,
        contrast_range=(0.5, 1.5),
        saturation_range=(0.5, 1.5),
        hue_delta=18),
    dict(
        type='Expand',
        mean=img_norm_cfg['mean'],
        to_rgb=img_norm_cfg['to_rgb'],
        ratio_range=(1, 4)),
    dict(
        type='MinIoURandomCrop',
        min_ious=(0.1, 0.3, 0.5, 0.7, 0.9),
        min_crop_size=0.3),
    dict(type='Resize', img_scale=(img_size, img_size), keep_ratio=False),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels', 'gt_masks']),
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(img_size, img_size),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=False),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img']),
        ])
]
data = dict(
    samples_per_gpu=8,
    workers_per_gpu=4,
    train=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/train.json',
        img_prefix=data_root + 'train/image/',
        pipeline=train_pipeline),
    val=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/val.json',
        img_prefix=data_root + 'val/image/',
        pipeline=test_pipeline),
    test=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/test.json',
        img_prefix=data_root + 'test/image/',
        pipeline=test_pipeline))
optimizer = dict(type='SGD', lr=1e-3, momentum=0.9, weight_decay=5e-4)
optimizer_config = dict()
# learning policy
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=500,
    warmup_ratio=0.1,
    step=[8, 11])
total_epochs = 10
cudnn_benchmark = True
evaluation = dict(metric=['bbox', 'segm'])
checkpoint_config = dict(interval=1)
work_dir = './logs_pig/yolact_r50/albu_new'
```

**Comment by jshilong** [comment]
All the model in the model zoo has been removed the redundant information such as `optimizer`.

**Comment by zhiweihu1103** [comment]
> All the models in the model zoo have been removed the redundant information such as `optimizer`.

Are you talking about calculation in FPS? I made the following modifications to the /mmdet/models/detectors/yolact.py file to remove some initialization and post-processing operation time, I don’t know if it is correct.
```
ef simple_test(self, img, img_metas, rescale=False):
        """Test function without test time augmentation."""
        total_time = 0

        x = self.extract_feat(img)

        torch.cuda.synchronize()
        start_time = time.perf_counter()
        cls_score, bbox_pred, coeff_pred = self.bbox_head(x)

        bbox_inputs = (cls_score, bbox_pred,
                       coeff_pred) + (img_metas, self.test_cfg, rescale)
        det_bboxes, det_labels, det_coeffs = self.bbox_head.get_bboxes(
            *bbox_inputs)

        bbox_results = [
            bbox2result(det_bbox, det_label, self.bbox_head.num_classes)
            for det_bbox, det_label in zip(det_bboxes, det_labels)
        ]
        num_imgs = len(img_metas)
        scale_factors = tuple(meta['scale_factor'] for meta in img_metas)
        if all(det_bbox.shape[0] == 0 for det_bbox in det_bboxes):
            segm_results = [[[] for _ in range(self.mask_head.num_classes)]
                            for _ in range(num_imgs)]
        else:
            # if det_bboxes is rescaled to the original image size, we need to
            # rescale it back to the testing scale to obtain RoIs.
            if rescale and not isinstance(scale_factors[0], float):
                scale_factors = [
                    torch.from_numpy(scale_factor).to(det_bboxes[0].device)
                    for scale_factor in scale_factors
                ]
            _bboxes = [
                det_bboxes[i][:, :4] *
                scale_factors[i] if rescale else det_bboxes[i][:, :4]
                for i in range(len(det_bboxes))
            ]

            mask_preds = self.mask_head(x[0], det_coeffs, _bboxes, img_metas)
            torch.cuda.synchronize()
            total_time += time.perf_counter() - start_time

            # apply mask post-processing to each image individually
            segm_results = []
            for i in range(num_imgs):
                if det_bboxes[i].shape[0] == 0:
                    segm_results.append(
                        [[] for _ in range(self.mask_head.num_classes)])
                else:
                    segm_result = self.mask_head.get_seg_masks(
                        mask_preds[i], det_labels[i], img_metas[i], rescale)
                    segm_results.append(segm_result)

        # return list(zip(bbox_results, segm_results))
        return list(zip(bbox_results, segm_results)), total_time
```

**Comment by zhiweihu1103** [comment]
> All the models in the model zoo have been removed the redundant information such as `optimizer`.

Another problem is why the training speed of YPLACT is slower than Mask RCNN. Thx!

**Comment by zhiweihu1103** [comment]
> All the models in the model zoo have been removed the redundant information such as `optimizer`.

Another problem is why the training speed of YPLACT is slower than Mask RCNN. Thx!

**Comment by jshilong** [comment]
The contributor of `yolact` in mmdetection adds a comment in the readme. 
[readme of yolact](https://github.com/open-mmlab/mmdetection/tree/master/configs/yolac)

*Note: The FPS is evaluated by the original implementation. When calculating FPS, only the model inference time is taken into account. Data loading and post-processing operations such as converting masks to RLE code, generating COCO JSON results, 
image rendering is not included.

We have not benchmarked the actual speed of the version provided by the author in mmdet.

**Comment by zhiweihu1103** [comment]
> The contributor of `yolact` in mmdetection adds a comment in the readme.
> [readme of yolact](https://github.com/open-mmlab/mmdetection/tree/master/configs/yolact)
> 
> *Note: The FPS is evaluated by the original implementation. When calculating FPS, only the model inference time is taken into account. Data loading and post-processing operations such as converting masks to RLE code, generating COCO JSON results,
> image rendering is not included.
> 
> We have not benchmarked the actual speed of the version provided by the author in mmdet.

Thx!
